src/main/riddl/Venues/Reservations/reservation.riddl

```diff
@@ -0,0 +1,115 @@
+		            }
+		            on command EditReservationInfo {
+		                then "set Draft info fields to EditReservationInfo.info"
+		                and "yield event ReservationInfoUpdated"
+		            }
+		            on command UpdateReservationStatus {
+                    then "morph Reservation to the state provided in UpdateReservationStatus"
+                    and "set MetaInfo to capture change data"
+                    and "yield event ReservationStatusUpdated"		            }
+		        }
+		    }
+
+		    state OnHold is {
+		        fields {
+		            reservationId: ReservationId,
+		            reservationInfo: ReservationInfo,
+		            reservationMeta: ReservationContext.MetaInfo
+		        }
+		        handler OnHoldReservation is {
+		            on command EstablishReservation {
+		                then error "Reservation already established"
+		            }
+		            on command EditReservationInfo {
+		                then "set OnHold info fields to EditReservationInfo.info"
+		                and "yield event ReservationInfoUpdated"
+		            }
+		            on command UpdateReservationStatus {
+                    then "morph Reservation to the state provided in UpdateReservationStatus"
+                    and "set MetaInfo to capture change data"
+                    and "yield event ReservationStatusUpdated"			            }
+		        }
+		    }
+
+		    state Reserved is {
+		        fields {
+		            reservationId: ReservationId,
+		            reservationInfo: ReservationInfo,
+		            attendees: MemberId*,
+		            reservationMeta: ReservationContext.MetaInfo
+		        }
+		        handler ReservedReservation is {
+		            on command EstablishReservation {
+		                then error "Reservation already established"
+		            }
+		            on command EditReservationInfo {
+		                then error "Reservation is in Reserved state, editing info not allows"
+		            }
+		            on command UpdateReservationStatus {
+                    then "morph Reservation to the state provided in UpdateReservationStatus"
+                    and "set MetaInfo to capture change data"
+                    and "yield event ReservationStatusUpdated"			            }
+		        }
+		    }
+
+		    state Active is {
+		        fields {
+		            reservationId: ReservationId,
+		            reservationInfo: ReservationInfo,
+		            attendees: MemberId+,
+		            reservationMeta: ReservationContext.MetaInfo
+		        }
+		        handler ActiveReservation is {
+		            on command EstablishReservation {
+		                then error "Reservation already established"
+		            }
+		            on command EditReservationInfo {
+		                then error "Reservation is in Active state, editing info not allowed"
+		            }
+		            on command UpdateReservationStatus {
```

Comment:
I think the only state we could transition to from Active is Past, correct? We might want to list examples. See organization.riddl for examples. Briefly here:
state Active is {
  handler ActiveReservation is {
    on command UpdateReservationStatus {
      example NewStatusIsDraft is {
        given "Reservation current status is Active"
        when "newStatus is Draft"
        then "error: Active reservations cannot transition back to a Draft state"
      }
   }
}

Reply:
An interesting thing I want to point out: if we are doing this, we need to have a gateway that is always conscious of who initiated a cascading command. Say this command is issued as an effect of another command, and we have to return the error to the user. Or better (or worse) yet, take the situation with Location being master over the VenueLocations projection. If a user makes a new Location with an invalid VenueId, we have to listen for these errors coming back and return them to user for clarity.

This also begs the question - how are we sending automatic updates? Updating projections automatically by listening for messages is a PubSub/Kafka thing, not a GRPC thing.

Reply:
Messaging would be done using PubSub/Kafka/NATS/Pulsar/.....(some other similar thing). And you are correct on the error thing. With Async messaging between components the caller would have no other choice but to listen for success/failure. Architecture is all about tradeoffs. You want scale, performance, resilience, reliability, etc.? You need to figure out how to rethink your UX and how to handle communications. If you are not so worried about those things synchronous communication may be the right choice.

Reply:
We could create a projection in the Gateway service that facilitates all of that. It would be pretty complex, but we can build it so it isn't a bottleneck. The real bottleneck is that we would have to rely on Firebase notifications or something similar to get to the actual device that issued the command.

Reply:
Actually, we would also have to undo commands that led to these events. We don't want Location with bad VenueId in the Location service, would be better to wipe it before sending the error back to client, since the operation was ultimately not a success.

Reply:
I have a different understanding of event sourcing. As I understand it, event sourcing is the practice of maintaining a log of the events that happen in a system. This log is the authoritative source of the state of an entity. In a "traditional" system the source is the current state. For example, If I have an entity Account with an attribute balance, and that balance is $68.23 that is all I know.

In an event sourced system I have a log of events:
id.      Action.      Amount.     Balance
1234 Deposited $100.00     $100.00
1235 Deposited $33.87       $133.87
1236 Withdrew  $53.27       $  80.60
1237 Deposited $25.00       $105.60
1238 Withdrew  $37.37        $ 68.23

I think either the front end or the gateway (more likely) would need to subscribe to the event stream to listen for events and reply to the device/client.

Reply:
Let's say in your example that one of the `Withdraw` events was initiated as a mistake - the wrong account was specified, and it withdrew from this account instead of some other account. You would have to issue another action to deposit that money back into the account by changing order payment information before shipping goes out. The system recognizes that the payment information changed and initiates 2 actions - `Withdraw` to new account and `Deposit` to old account.

The same would apply to creating a new Location with a VenueId that does not exist. Furthermore, Gateway will have no way to associate the `ActionID` of the `UpdateVenueWithLocation` action with that of the `CreateLocation` action. Granted, it can search through the event logs for the `CreateLocation` that contains the  `LocationId` specified in `UpdateVenueWithLocation` . But I don't there's there's any assurance we will always be have enough of an identifying factor.

Reply:
We may not be using the best example. When `UpdateVenueWithLocation` fails we can have Gateway ask App for a different `VenueId` and send another `UpdateVenueWithLocation` as well as an `UpdateLocationWithVenue`, or else they can delete their new `Location`.

It does come into play though for the automatic availability query I'm writing. They are very much up in the air right now (see my comments here https://github.com/improving-app/riddl/pull/23), but it is necessarily a multistep operation where we ask for availability from a projection that joins `Location`s to their `Reservation`s, and then go to `ReservationContext` to update the state of the randomly-chosen-from-search-results `ReservationId`... The success message returned by `ReservationContext` will need to use its `LocationId` and `ReservationId` and `timestamp` to find the originating `User` in the logs...

So never mind about this :) interesting stuff to think about, but not really an issue